File: src/com/company/App/Expression.java
#pragma once

#include "IExpression.h"
#include "../Commands/ICommand.h"
#include "../Commands/IDisplayable.h"

namespace Apps {
	class Expression : public Commands::ICommand, public IExpression, public Commands::IDisplayable {
	public:
		/*********************************************************************************
		* Base class for all expression
		**********************************************************************************
		*/
		Expression();
		/*********************************************************************************
		* Get list of tokens for this expression
		**********************************************************************************
		*/
		const vector<Token>& GetSymbol() const;
		/*********************************************************************************
		* Return string to be displayed
		**********************************************************************************
		*/
		virtual string DisplayString() const;
		/*********************************************************************************
		* Return name as a command
		**********************************************************************************
		*/
		string GetCommandName() const;
		/*********************************************************************************
		* Get the original input
		**********************************************************************************
		*/
		string GetInput() const;
		/*********************************************************************************
		* Destructor
		**********************************************************************************
		*/
		virtual ~Expression();
	protected:
		vector<Token> m_symbols;
		string m_commandName;
		string m_input;
	};
}

File: src/com/company/App/src/Expression.cpp
#include "../stdafx.h"
#include "../Expression.h"

namespace Apps {
	Expression::Expression() {}
	Expression::~Expression() {}
	const vector<Token>& Expression::GetSymbol() const {
		return m_symbols;
	}
	string Expression::DisplayString() const {
		Token::NumeralType numType;
		for (auto const& i: m_symbols) {
			if (i.GetNumeralType() != Token::NumeralType::NUM_UNDEFINED)
				numType = i.GetNumeralType();
		}
		string m_converted;
		try {
			m_converted = Token::ToNumeralType(Calculate(), numType);
		}
		catch (CantConvertException ex) {
			cerr << "Tidak dapat mengembalikan hasil perhitungan dalam mode bilangan ini" << endl;
		}
		return m_input + " = " + m_converted;
	}
	string Expression::GetInput() const {
		return m_input;
	}
	string Expression::GetCommandName() const {
		return m_commandName;
	}
}
